File: Leetcode/0108 Convert Sorted Array to Binary Search Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        return sortedArrayToBST(0,nums.size() - 1,nums);
    }
    
    TreeNode* sortedArrayToBST(int left,int right,vector<int>& nums) {
        if(left > right)
            return NULL;
        
        int mid = (left + right) >> 1;
        
        TreeNode* root = new TreeNode(nums[mid]);
        
        root->left  = sortedArrayToBST(left   , mid - 1, nums);
        root->right = sortedArrayToBST(mid + 1, right  , nums);
        
        return root;
    }
};

File: Leetcode/0108 Convert Sorted Array to Binary Search Tree.ts
function sortedArrayToBST(nums: number[]): TreeNode | null {
    return sortedArrayToBst(0, nums.length - 1, nums);
}

function sortedArrayToBst(left: number, right: number, nums: number[]): TreeNode | null {
    if (left > right) {
        return null;
    }
    const mid = left + Math.floor((right - left) / 2);
    const root = new TreeNode(nums[mid]);
    root.left = sortedArrayToBst(left, mid - 1, nums);
    root.right = sortedArrayToBst(mid + 1, right, nums);
    return root;
}
